quil-rs/src/program/frame.rs

```diff
@@ -154,3 +175,33 @@ pub(crate) enum FrameMatchCondition<'a> {
     /// Return all frames which match any of these conditions
     Or(Vec<FrameMatchCondition<'a>>),
 }
+
+/// A pair of conditions to match frames within a [`crate::Program`] (or another scope).
+///
+/// This allows for deferred evaluation of matching an instruction against available frames.
+pub(crate) struct FrameMatchConditions<'a> {
+    /// A condition to identify which frames within a [`crate::Program`] (or another scope)
+    /// are actively used by an [`Instruction`].
+    ///
+    /// If `None`, then this [`Instruction`] does not use any frames, regardless of which are available.
+    pub used: Option<FrameMatchCondition<'a>>,
+
+    /// A condition to identify which frames within a [`crate::Program`] (or another scope)
+    /// are blocked by an [`Instruction`]. A "blocked" frame is one which is not used by the
+    /// `Instruction` but is not available for use by other instructions while this one executes.
+    ///
+    /// **Note**: for efficiency in computation, this may match frames also matched by `used`.
+    /// In order to query which frames are _blocked but not used_, both conditions must first
+    /// be evaluated in the scope of the available frames.
+    pub blocked: Option<FrameMatchCondition<'a>>,
+}
+
+/// The product of evaluating  [`FrameMatchConditions`] in the scope of available frames (such as within a [`crate::Program`]).
+pub struct MatchedFrames<'a> {
+    /// Which concrete frames are used by the [`Instruction`]
+    pub used: Option<HashSet<&'a FrameIdentifier>>,
+
+    /// Which concrete frames are blocked and not used.
+    /// This set must be mutually exclusive with `used`.
+    pub blocked: Option<HashSet<&'a FrameIdentifier>>,
+}
```

Comment:
In order to ensure that the mutual exclusivity is upheld, should these fields be private with immutable getters?

Alternatively, would it make more sense to have a single `HashMap<&'a FrameIdentifier, FrameMatchType>` where

```rust
enum FrameMatchType {
    Borrowed,
    Used,
}
```

(I like the separation of these fields, I'm just looking at how we can ensure the exclusivity is always true)

Reply:
> should these fields be private with immutable getters?

Yes, that's a good idea - since this struct is meant to be derived in an analysis step and then remain immutable.

> would it make more sense to have a single

In practice, these two groups are iterated over, which would become less efficient if we had to iterate over both with mutually exclusive filters for each.